PyGA/GA.py

```diff
@@ -0,0 +1,605 @@
+                self.save_to_tracker()
+                self.evaluate_population()
+                
+        if self.settings["save_all_files"]:
+            self.save_population()
+            self.write_population_genes()
+            #self.write_population_fitness()
+            
+
+    def print_stats(self):
+        print(f"Generation: {self.generation}")
+        print(f"Best Fitness: {self.best_fitness}")
+        print(f"Best Individual: {self.best_individual.genes}")
+    
+    
+    ### Generational Methods #################################################
+    
+    def evaluate_population(self):
+        ''' Evaluate the fitness of the entire population'''
+        for individual in self.population:
+            individual.evaluate_fitness(self.comparison)
+            if individual.fitness > self.best_fitness:
+                self.best_fitness = individual.fitness
+                self.best_individual = individual
+    
+    
+    def get_operator_numbers(self):
+        mutation_no = int(self.settings["mutation_rate"] * self.settings["npop"])
+        crossover_no = int(self.settings["crossover_rate"] * self.settings["npop"])
+        reproduction_no = int(self.settings["reproduction_rate"] * self.settings["npop"])
+        
+        if crossover_no % 2 != 0:
+            crossover_no += 1
+        
+        if mutation_no + crossover_no + reproduction_no < self.settings["npop"]:
+            injection_no = self.settings["npop"] - mutation_no - crossover_no - reproduction_no
+        else:
+            injection_no = 0
+            reproduction_no = self.settings["npop"] - mutation_no - crossover_no
+            
+        return [crossover_no, mutation_no, reproduction_no, injection_no]
+    
+    
+    ### SSGA Methods #########################################################
+    
+    def choose_operator(self):
+        '''choose an operator from the operator set of 
+        REPRODUCTION, CROSSOVER, MUTATION, INJECTION'''
+        # choose a random number from 0 to 1
+        choice = random.uniform(0, 1)
+        
+        limit = self.settings["crossover_rate"]
+        if choice <= limit:
+            return "crossover"
+        
+        limit += self.settings["mutation_rate"]
+        if choice <= limit:
+            return "mutation"
+        
+        limit += self.settings["reproduction_rate"]
+        if choice <= limit:
+            return "reproduction"
+        
+        return "injection"    
+
+    
+    def create_individual(self, operator, parents):
+        '''Create an individual from the operator and parents'''
+        
+        if operator == "crossover":
+            new_indiv = self.crossover(parents[0], parents[1])
+        elif operator == "mutation":
+            new_indiv = self.mutation(parents[0])
+        elif operator == "injection":
+            new_indiv = self.injection()
+        else:
+            new_indiv = parents[0]
+        
+        if type(new_indiv) in [list, tuple]:
```

Comment:
I originally did it because I didn't know if the crossover function would return a list or a tuple, and was going to check later. I'm actually just changing new_indiv to grab the first individual from crossover and getting rid of the type checking